### ISSUE TITLE
Segmentation fault (stack overflow) due to unbounded recursion

### ISSUE THREAD

**Issue by guidovranken** [nlohmann/json#1419]
The following causes a segmentation fault (stack overflow) due to an apparent unbounded recursion.

```cpp
#include <string>
#include "json.hpp"
int main(void)
{
    const std::string s(512000, '[');
    try {
        auto j = nlohmann::json::parse(s);
    } catch ( ... )  { } ;

    return 0;
}
```

**Comment by nickaein** [comment]
Good catch! 

There is also a similar issue related to `write_ubjson` recursive implementation: https://github.com/nlohmann/json/pull/1414

The example you posted demonstrate another case of recursion. While I couldn't find an obvious recursive implementation in `parser` and `json_sax`, the behavior tells a case of stack smashing.

Re-writing an recursive implementation as a non-recursive is generally an improvement. However, there might be some considerations in revising these algorithms since they are fundamental part of the library. @nlohmann can give some insight on how to proceed.

**Comment by nickaein** [comment]
The recursive happens during array deconstruction [here](https://github.com/nlohmann/json/blob/db53bdac1926d1baebcb459b685dcd2e4608c355/include/nlohmann/json.hpp#L1010). The saniziter

**Comment by nlohmann** [comment]
Thanks for reporting!

The parser itself is not recursive - the stack is modeled by a `std::vector<bool>` as we only need to know the nesting of objects and arrays. The input itself is not the problem, but the way the library models values. The input opens and nests one `std::vector<basic_json>` with each opening `[`.

It would be great if we could find a better way for this. Or a way to detect the issue and throw an exception before it happens.

The issue was already detected by Google's OSS Fuzz, see #832. There, I already asked for some opinions on how to proceed. I would mark this issue as duplicate and invite for a discussion in #832.

**Comment by nlohmann** [comment]
Fixed by merging #1436.
